File: python/questions/nextPermutation.py
from typing import List

class Solution:

    def fb(self, l: List[int], b: int, e: int, t: int) -> int:
        while b < e:
            m = (b + e + 1) // 2
            if l[m] > t:
                b = m
            else:
                e = m - 1
        return b

    def nextPermutation(self, nums: List[int]) -> None:
        ln = len(nums)
        p = ln - 1
        while p>0 and nums[p]<=nums[p-1]:
            p -= 1
        if p:
            bp = self.fb(nums, p, ln-1, nums[p-1])
            nums[p-1], nums[bp] = nums[bp], nums[p-1]
        l = p
        r = ln - 1
        while l < r:
            nums[l], nums[r] = nums[r], nums[l]
            l += 1
            r -= 1

File: cpp/questions/nextPermutation.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
private:
    int fb(std::vector<int>& v, int l, int r, int t) {
        while (l < r) {
            int m = (l + r + 1) / 2;
            if (v[m] > t) {
                l = m;
            } else {
                r = m - 1;
            }
        }
        return l;
    }
public:
    void nextPermutation(vector<int>& nums) {
        int ns = nums.size();
        int p = ns - 1;
        for (; p>0; --p) {
            if (nums[p] > nums[p-1]) {
                break;
            }
        }
        if (p != 0) {
            int bp = fb(nums, p, ns-1, nums[p-1]);
            std::swap(nums[p-1], nums[bp]);
        }
        int s = p;
        int e = ns - 1;
        while (s < e) {
            std::swap(nums[s++], nums[e--]);
        }
    }
};
